x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexing/AsyncTwoPhaseIndexer.java

```diff
@@ -148,16 +145,18 @@ public synchronized boolean maybeTriggerAsyncJob(long now) {
         case STARTED:
             logger.debug("Schedule was triggered for job [" + getJobId() + "], state: [" + currentState + "]");
             stats.incrementNumInvocations(1);
-            onStartJob(now);
 
             if (state.compareAndSet(IndexerState.STARTED, IndexerState.INDEXING)) {
                 // fire off the search. Note this is async, the method will return from here
                 executor.execute(() -> {
                     try {
-                        stats.markStartSearch();
-                        doNextSearch(buildSearchRequest(), ActionListener.wrap(this::onSearchResponse, this::finishWithSearchFailure));
+                        onStart(now, () -> {
+                            stats.markStartSearch();
+                            doNextSearch(buildSearchRequest(), ActionListener.wrap(this::onSearchResponse, this::finishWithSearchFailure));
+                        });
                     } catch (Exception e) {
```

Comment:
I don't think we should catch the exception here, if an error occurs in `onStart` we should provide an ActionListener to let the caller deals with the error (e.g. call listener.onFailure(e)):
````
onStart(now,  ActionListener.wrap((o) -> {
                            stats.markStartSearch();
                            doNextSearch(buildSearchRequest(), ActionListener.wrap(this::onSearchResponse, this::finishWithSearchFailure));
                        }, (e) -> doSaveState(finishAndSetState(), position.get(), () -> onFailure(e));
````

Reply:
I also wonder if we should call onFailure before we save the state to be consistent with onFinish : `onFailure(Exception e, Runnable finishAndSaveState)` ?